internal/trace/transport.go

```diff
@@ -0,0 +1,74 @@
+// Copied and adapted from oras (https://github.com/oras-project/oras)
+/*
+Copyright The ORAS Authors.
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+
+Copied and adapted from oras (https://github.com/oras-project/oras)
+*/
+
+package trace
+
+import (
+	"net/http"
+	"strings"
+
+	"github.com/notaryproject/notation-go/log"
+)
+
+// Transport is an http.RoundTripper that keeps track of the in-flight
+// request and add hooks to report HTTP tracing events.
+type Transport struct {
+	http.RoundTripper
+}
+
+func NewTransport(base http.RoundTripper) *Transport {
+	return &Transport{base}
+}
+
+// RoundTrip calls base roundtrip while keeping track of the current request.
+func (t *Transport) RoundTrip(req *http.Request) (resp *http.Response, err error) {
+	ctx := req.Context()
+	e := log.GetLogger(ctx)
+
+	e.Debugf(" Request URL: %q", req.URL)
+	e.Debugf(" Request method: %q", req.Method)
+	e.Debugf(" Request headers:")
+	logHeader(req.Header, e)
+
+	resp, err = t.RoundTripper.RoundTrip(req)
+	if err != nil {
+		e.Errorf("Error in getting response: %w", err)
+	} else if resp == nil {
+		e.Errorf("No response obtained for request %s %q", req.Method, req.URL)
+	} else {
+		e.Debugf(" Response Status: %q", resp.Status)
+		e.Debugf(" Response headers:")
+		logHeader(resp.Header, e)
+	}
+	return resp, err
+}
+
+// logHeader prints out the provided header keys and values, with auth header
+// scrubbed.
+func logHeader(header http.Header, e log.Logger) {
```

Comment:
Do we need to log all headers? Isn't this too much information for sign and verify commands, the [sample log](https://hackmd.io/SJ1IXiiOQESuuS9qQj0GPw) that you shared contained ~50% or more log lines that were just network calls information?

```
DEBU[2022-12-02T13:14:39+08:00]  Request headers:                            
DEBU[2022-12-02T13:14:39+08:00]    "Range": "bytes=0-2220"                   
DEBU[2022-12-02T13:14:39+08:00]    "User-Agent": "notation/v0.12.0-beta.1+unreleased" 
DEBU[2022-12-02T13:14:39+08:00]  Response Status: "206 Partial Content"      
DEBU[2022-12-02T13:14:39+08:00]  Response headers:                           
DEBU[2022-12-02T13:14:39+08:00]    "Content-Length": "2221"                  
DEBU[2022-12-02T13:14:39+08:00]    "Content-Range": "bytes 0-2220/2221"      
DEBU[2022-12-02T13:14:39+08:00]    "Docker-Distribution-Api-Version": "registry/2.0" 
DEBU[2022-12-02T13:14:39+08:00]    "X-Content-Type-Options": "nosniff"       
DEBU[2022-12-02T13:14:39+08:00]    "Date": "Fri, 02 Dec 2022 05:14:39 GMT"   
DEBU[2022-12-02T13:14:39+08:00]    "Accept-Ranges": "bytes"                  
DEBU[2022-12-02T13:14:39+08:00]    "Cache-Control": "max-age=31536000"       
DEBU[2022-12-02T13:14:39+08:00]    "Content-Type": "application/octet-stream" 
DEBU[2022-12-02T13:14:39+08:00]    "Docker-Content-Digest": "sha256:9e27c57b266d8bcd206a90af96dba94a6c2d9ac8fe93d47979aaf7ce47a34f68" 
DEBU[2022-12-02T13:14:39+08:00]    "Etag": "\"sha256:9e27c57b266d8bcd206a90af96dba94a6c2d9ac8fe93d47979aaf7ce47a34f68\"" 

```

Reply:
@priteshbandi I got an `403 Forbidden` error when I open the link. Could you copy it here or grant a permission.

Reply:
Yes, it is lengthy. However, the http header is requred for debuging issue related to registry and different registry has different headers. For debug log, usability is more important than beautiful degree. Can we keep them here for now?